8262438: sun/security/ssl/SSLLogger/LoggingFormatConsistency.java failed with "SocketException: Socket is closed"

evwhelan opened the pull request:
Hi all,

Please review my test fix relating to JDK-8262438

This patch introduces as Thread.sleep at the start of each iteration which creates a new test jvm. 
This allows the server socket sufficient time to release the previous connection and allows the port to be used again.

I also refactored the behaviour for when the exitCode is not 0, allowing for an easier to read output.
An incorrect HttpsUrlConnection.disconnect() was also removed.

The test was ran 100 times on all platforms and no failures were seen. 

Thanks,
Evan

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8262438](https://bugs.openjdk.java.net/browse/JDK-8262438): sun/security/ssl/SSLLogger/LoggingFormatConsistency.java failed with "SocketException: Socket is closed"


### Reviewers
 * [Rajan Halade](https://openjdk.java.net/census#rhalade) (@rhalade - **Reviewer**)


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2749/head:pull/2749`
`$ git checkout pull/2749`

dfuch commented:
Hi Evan - I am a bit skeptical that the proposed fix will solve the issue. AFAICS the exception is raised by the server side - and if I read it correctly it happens when the server finds that the socket is already closed when it tries to close it. How could sleeping between 2 process invocation solve this? The port used is an ephemeral port - so it should never be the same?
Does the log show that the same port was being reused?

evwhelan commented:
Hi Daniel, 

From speaking with you offline, I have ensured that the server-side and client-side socket input streams are emptied.

I have also added some debugging info should this test fail in the future.

I've removed the proposed `Thread.sleep` and the test does not fail in 400 runs

Regards,
Evan

dfuch commented on test/jdk/sun/security/ssl/SSLLogger/LoggingFormatConsistency.java:
```diff
@@ -117,10 +123,14 @@ protected void runServerApplication(SSLSocket socket) throws Exception {
             out.writeBytes("Content-Type: text/html\r\n\r\n");
             out.flush();
+        } finally {
+            socket.getInputStream().readAllBytes();
```
This will cause the server side to block until the client closes the socket. Is that what you really want to do? (It may be - but if the client is a regular HTTP client (HttpURLConnection / HttpClient) it will not close the connection until its keep-alive delay (may be up to 20mins) is expired.

evwhelan replied:
I've updated the PR by removing the custom server logic, more detail can be found in my latest comment.

Evan :)

evwhelan commented:
Hi Daniel,

So in testing, I realised that adding my own custom logic to handle the client-server communication was a bit overkill.

Simply invoking the `SSLSocketTemplate::run` method is sufficient in making the basic connection and generating the desired debug info which the test checks for.

After 400 runs I saw no failures

Thanks,
Evan

dfuch reviewed (commented):
I don't see any specific issue with what you are proposing from a networking perspective, but you will need someone from security-libs to review the proposed changes and acknowledge that the test is still testing what it intends to test.

robm-openjdk commented:
The changes here don't affect the original fix given that the default impl of the client & server work for the purposes of the test.

rhalade commented:
Can you please confirm that you saw all protocols ("TLSv1", "TLSv1.1", "TLSv1.2", "TLSv1.3") negotiated with updated test run?

evwhelan commented:
Yes, the updated test run successfully tested with all of the protocols listed.

evwhelan commented:
/integrate

robm-openjdk commented:
/sponsor
